README.md
# PerioDart
[![Pub](https://img.shields.io/pub/v/periodart.svg)](https://pub.dartlang.org/packages/periodart)
[![Build Status](https://travis-ci.org/emnljrz/periodart.svg?branch=master)](https://travis-ci.org/emnljrz/periodart)

Ez Dart DateTime Library :package: is still in development mode.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased] - [Future feature]
- Parsing
- Datetime format and display
- Minupulation function
- Basic datetime query
- More duration format

## [1.0.0] - 2018-05-29
### Added
- Added basic datetime function

test/test.dart
import 'package:test/test.dart';
import '../periodart.dart';

void main() {
  group('group test', () {
    String dateTimeString = "2018-05-31 12:45:00.123456";
    DateTime timeIn = DateTime.parse(dateTimeString);

    setUp(() {

    });

    print(timeIn);
    
    Period period = new Period(timeIn);

    test('Period Adjustment 1', () {
      period.Adjust(-15);
      period.ofMinutes();
      expect(period.Display() == DateTime.parse("2018-05-31 12:30:00.123456"), isTrue);
    });
  
    test('Period Adjustment 2', () {
      period.Adjust(15);
      period.ofMinutes();
      expect(period.Display() == DateTime.parse("2018-05-31 13:00:00.123456"), isTrue);
    });

    test('Period Adjustment 3', () {
      period.Adjust(6);
      period.ofMonths();
      expect(period.Display() == DateTime.parse("2018-11-30 12:45:00.123"), isTrue);
    });

    Fetch fetch = new Fetch(timeIn);

    test('Fetch 1', () {
      period.Adjust(15);
      period.ofMinutes();
      expect(fetch.TheDate() == "2018-05-31", isTrue);
    });

    test('Fetch 2', () {
      period.Adjust(15);
      period.ofMinutes();
      expect(fetch.TheMilitaryTimeWithMillisecond() == "12:45:00.123", isTrue);
    });

    test('Fetch 3', () {
      period.Adjust(15);
      period.ofMinutes();
      expect(fetch.TheRegularTime() == "12:45:00 PM", isTrue);
    });

    test('Fetch 4', () {
      period.Adjust(15);
      period.ofMinutes();
      expect(fetch.TheRegularTimeWithMillisecond() == "12:45:00.123 PM", isTrue);
    });

    test('Manipuldate 1', () {
      Manipulate manipulate = new Manipulate();
      manipulate.dateFrom  = timeIn;
      manipulate.dateTo    = DateTime.parse("2018-05-31 16:02:00");
      manipulate.WithPrefix();
      expect(manipulate.DisplayMoment() == "in 3 hours", isTrue);
    });

    test('Manipuldate 2', () {
      Manipulate manipulate = new Manipulate();
      manipulate.dateFrom  = DateTime.parse("2018-05-31 16:02:00");
      manipulate.dateTo    = timeIn;
      manipulate.WithPrefix();
      expect(manipulate.DisplayMoment() == "3 hours ago", isTrue);
    });

    test('Manipuldate 3', () {
      Manipulate manipulate = new Manipulate();
      manipulate.dateFrom  = DateTime.parse("2018-05-31 16:02:00");
      manipulate.dateTo    = timeIn;
      expect(manipulate.DisplayMoment() == "3 hours", isTrue);
    });

  });
}

src/Fetch.dart
import 'util.dart';

class Fetch {

  DateTime _baseValue;
  String _year, _month, _day;
  String _hour, _minute, _second, _mls;
  String _timePeriod;
  Util _util;


  Fetch(DateTime dateTimeValue) { 
    _util = new Util();
  	_baseValue = dateTimeValue;
    _year   = _util.paddedFormat(_baseValue.year);
    _month  = _util.paddedFormat(_baseValue.month);
    _day    = _util.paddedFormat(_baseValue.day);
    _hour   = _util.paddedFormat(_baseValue.hour);
    _minute = _util.paddedFormat(_baseValue.minute);
    _second = _util.paddedFormat(_baseValue.second);
    _mls    = _util.paddedFormat(_baseValue.millisecond);
  }

  TheDate() {
    return _year + "-" + _month  + "-" + _day;
  }

  TheMilitaryTime() {
    return _hour + ":" + _minute  + ":" + _second;
  }

  TheMilitaryTimeWithMillisecond() {
    return _hour + ":" + _minute + ":" + _second + "." + _mls;
  }

  TheRegularTime() {
    _timePeriod = int.parse(_hour) >= 12 && int.parse(_minute) > 0 ? "PM" : "AM";
    var nHour = int.parse(_hour) > 12 ? int.parse(_hour) - 12 : _hour;
    return  nHour.toString() + ":" + _minute.toString() + ":" + _second.toString() + " " + _timePeriod;
  }

  TheRegularTimeWithMillisecond() {
    _timePeriod = int.parse(_hour) >= 12 && int.parse(_minute) > 0  ? "PM" : "AM";
    var nHour = int.parse(_hour) > 12 ? int.parse(_hour) - 12 : _hour;
    return nHour.toString() + ":" + _minute.toString() + ":" + _second.toString() + "." + _mls + " " + _timePeriod;
  }


}


src/Manipulate.dart

class Manipulate {

  DateTime dateFrom;
  DateTime dateTo;
  bool _prefix = false;

  Manipulate() {
   
  }

  WithPrefix() { _prefix = true; }

  DisplayMoment() {
    Duration timeDiff = dateTo.difference(dateFrom);
    String newTime = "";

    if      (timeDiff.inSeconds.abs() < 45) newTime = "a few moments";
    else if (timeDiff.inMinutes.abs() < 2)  newTime = "a minute";
    else if (timeDiff.inMinutes.abs() < 45) newTime = "${timeDiff.inMinutes.abs()} minutes";
    else if (timeDiff.inHours.abs() < 2)    newTime = "an hour";
    else if (timeDiff.inHours.abs() < 22)   newTime = "${timeDiff.inHours.abs()} hours";
    else if (timeDiff.inDays.abs() < 2)     newTime = "a day";
    else if (timeDiff.inDays.abs() < 26)    newTime = "${timeDiff.inDays.abs()} days";
    else if (timeDiff.inDays.abs() < 60)    newTime = "a month";
    else if (timeDiff.inDays.abs() < 320)   newTime = "${timeDiff.inDays.abs() ~/ 30} months";
    else if (timeDiff.inDays.abs() < 547)   newTime = "a year";
    else                                    newTime = "${timeDiff.inDays.abs() ~/ 356} years";

    if (_prefix) {
      if (timeDiff.isNegative)
        newTime += " ago";
      else
        newTime = "in " + newTime;
    }
    return newTime;
  }

}


src/Period.dart
import 'util.dart';

class Period {
  
  DateTime _baseValue;
  DateTime _newValue;
  String _year, _month, _day;
  String _hour, _minute, _second, _mls;
  int _periodValue = 0;
  Util _util;

  Period(DateTime dateTimeValue) { 
  	_util = new Util();
  	_baseValue = dateTimeValue;
    _year   = _util.paddedFormat(_baseValue.year);
    _month  = _util.paddedFormat(_baseValue.month);
    _day    = _util.paddedFormat(_baseValue.day);
    _hour   = _util.paddedFormat(_baseValue.hour);
    _minute = _util.paddedFormat(_baseValue.minute);
    _second = _util.paddedFormat(_baseValue.second);
    _mls    = _util.paddedFormat(_baseValue.millisecond);
  }

  Now() { return new DateTime.now(); }

  Adjust(int value) {
    _periodValue = value;
  }

  Display() {
    return _newValue;
  }
  
  ofSeconds() {
    Duration duration = new Duration(seconds: _periodValue);
    _newValue = _baseValue.add(duration);
  }

  ofMinutes() {
    Duration duration = new Duration(minutes: _periodValue);
    _newValue = _baseValue.add(duration);
  }

  ofHours() {
    Duration duration = new Duration(hours: _periodValue);
    _newValue = _baseValue.add(duration);
  }
  
  ofDays() {
    Duration duration = new Duration(days: _periodValue);
    _newValue = _baseValue.add(duration);
  }

  ofMonths() {
    _year   = _util.paddedFormat(_baseValue.year);
    _month  = _util.paddedFormat(_baseValue.month + _periodValue);
    _day    = _baseValue.day <= 30 ? _util.paddedFormat(_baseValue.day) : _util.paddedFormat(endOfTheMonth((_baseValue.month + _periodValue)));

    _newValue = DateTime.parse(_year + "-" + _month + "-" + _day + " " + _hour + ":" + _minute + ":" + _second + "." + _mls);
  }
  
  ofYears() {
    _year   = _util.paddedFormat(_baseValue.year  + _periodValue);
    _month  = _util.paddedFormat(_baseValue.month);
    _day    = _util.paddedFormat(_baseValue.day);

    _newValue = DateTime.parse(_year + "-" + _month + "-" + _day + " " + _hour + ":" + _minute + ":" + _second + "." + _mls);
  }
  
  isMonthWith31(Object value) {
    return endOfTheMonth(value) == 31 ? true : false;
  }

  endOfTheMonth(Object value) {
    int month = value is DateTime ? value.month : value;
    int maxDate;

    if (month == 1) maxDate = 31;
    else if (month == 2 && !isLeapYear(value)) maxDate = 28;
    else if (month == 2 && isLeapYear(value)) maxDate = 29;
    else if (month == 3) maxDate = 31;
    else if (month == 4) maxDate = 30;
    else if (month == 5) maxDate = 31;
    else if (month == 6) maxDate = 30;
    else if (month == 7) maxDate = 31;
    else if (month == 8) maxDate = 31;
    else if (month == 9) maxDate = 30;
    else if (month == 10) maxDate = 31;
    else if (month == 11) maxDate = 30;
    else if (month == 12) maxDate = 31;

    return maxDate;
  }

  isLeapYear(Object value) {
      int year = value is DateTime ? value.year : value;

      if (year % 4 == 0) return true;
      else if (year % 400 == 0) return true;
      else if (year % 100 == 0) return false;
      else return false;
  }

}
